File: Neocom/NCCharacterAttributes.m
//
//  NCCharacterAttributes.m
//  Neocom
//
//  Created by Артем Шиманский on 14.01.14.
//  Copyright (c) 2014 Artem Shimanski. All rights reserved.
//

#import "NCCharacterAttributes.h"
#import "EVEDBAPI.h"
#import "EVEOnlineAPI.h"

@interface NCCharacterAttributes ()
- (NSInteger) effectiveAttributeValueWithAttributeID:(NSInteger) attributeID;
@end

@implementation NCCharacterAttributes

+ (instancetype) defaultCharacterAttributes {
	NCCharacterAttributes* attributes = [[NCCharacterAttributes alloc] init];
	attributes.charisma = 19;
	attributes.intelligence = 20;
	attributes.memory = 20;
	attributes.perception = 20;
	attributes.willpower = 20;
	return attributes;
}

- (id) initWithCharacterSheet:(EVECharacterSheet*) characterSheet {
	if (self = [super init]) {
		if (characterSheet) {
			self.intelligence = characterSheet.attributes.intelligence;
			
			self.charisma = characterSheet.attributes.charisma;
			self.intelligence = characterSheet.attributes.intelligence;
			self.memory = characterSheet.attributes.memory;
			self.perception = characterSheet.attributes.perception;
			self.willpower = characterSheet.attributes.willpower;
			
			for (EVECharacterSheetAttributeEnhancer *enhancer in characterSheet.attributeEnhancers) {
				switch (enhancer.attribute) {
					case EVECharacterAttributeCharisma:
						self.charisma += enhancer.augmentatorValue;
						break;
					case EVECharacterAttributeIntelligence:
						self.intelligence += enhancer.augmentatorValue;
						break;
					case EVECharacterAttributeMemory:
						self.memory += enhancer.augmentatorValue;
						break;
					case EVECharacterAttributePerception:
						self.perception += enhancer.augmentatorValue;
						break;
					case EVECharacterAttributeWillpower:
						self.willpower += enhancer.augmentatorValue;
						break;
				}
			}
		}
		else {
			self.charisma = 19;
			self.intelligence = 20;
			self.memory = 20;
			self.perception = 20;
			self.willpower = 20;
		}
	}
	return self;
}

- (float) skillpointsPerSecondForSkill:(EVEDBInvType*) skill {
	EVEDBDgmTypeAttribute *primaryAttribute = skill.attributesDictionary[@(180)];
	EVEDBDgmTypeAttribute *secondaryAttribute = skill.attributesDictionary[@(181)];
	NSInteger effectivePrimaryAttribute = [self effectiveAttributeValueWithAttributeID:primaryAttribute.value];
	NSInteger effectiveSecondaryAttribute = [self effectiveAttributeValueWithAttributeID:secondaryAttribute.value];
	return (effectivePrimaryAttribute + effectiveSecondaryAttribute / 2.0) / 60.0;
}

#pragma mark - NSCoding

- (void) encodeWithCoder:(NSCoder *)aCoder {
	[aCoder encodeInteger:self.charisma forKey:@"charisma"];
	[aCoder encodeInteger:self.intelligence forKey:@"intelligence"];
	[aCoder encodeInteger:self.memory forKey:@"memory"];
	[aCoder encodeInteger:self.perception forKey:@"perception"];
	[aCoder encodeInteger:self.willpower forKey:@"willpower"];
}

- (id) initWithCoder:(NSCoder *)aDecoder {
	if (self = [super init]) {
		self.charisma = [aDecoder decodeIntegerForKey:@"charisma"];
		self.intelligence = [aDecoder decodeIntegerForKey:@"intelligence"];
		self.memory = [aDecoder decodeIntegerForKey:@"memory"];
		self.perception = [aDecoder decodeIntegerForKey:@"perception"];
		self.willpower = [aDecoder decodeIntegerForKey:@"willpower"];
	}
	return self;
}

#pragma mark - Private

- (NSInteger) effectiveAttributeValueWithAttributeID:(NSInteger) attributeID {
	switch (attributeID) {
		case 164:
			return self.charisma;
			break;
		case 165:
			return self.intelligence;
			break;
		case 166:
			return self.memory;
			break;
		case 167:
			return self.perception;
			break;
		case 168:
			return self.willpower;
			break;
		default:
			break;
	}
	return 0;
}

@end

File: Neocom/Neocom/NCCharacterAttributes.swift
//
//  NCCharacterAttributes.swift
//  Neocom
//
//  Created by Artem Shimanski on 02.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import Foundation
import EVEAPI

class NCCharacterAttributes {
	var intelligence: Int = 20
	var memory: Int = 20
	var perception: Int = 20
	var willpower: Int = 20
	var charisma: Int = 19
	
	init() {
	}
	
	init(characterSheet: EVECharacterSheet) {
		NCDatabase.sharedDatabase?.performTaskAndWait({ (managedObjectContext) in
			let invTypes = NCDBInvType.invTypes(managedObjectContext: managedObjectContext)
			for implant in characterSheet.implants {
				if let attributes = invTypes[implant.typeID]?.allAttributes {
					self.intelligence += Int(attributes[NCDBAttributeID.intelligenceBonus.rawValue]?.value ?? 0)
					self.memory += Int(attributes[NCDBAttributeID.memoryBonus.rawValue]?.value ?? 0)
					self.perception += Int(attributes[NCDBAttributeID.perceptionBonus.rawValue]?.value ?? 0)
					self.willpower += Int(attributes[NCDBAttributeID.willpowerBonus.rawValue]?.value ?? 0)
					self.charisma += Int(attributes[NCDBAttributeID.charismaBonus.rawValue]?.value ?? 0)
				}
			}
		})
	}
	
	func skillpointsPerSecond(forSkill skill: NCSkill) -> Double {
		return skillpointsPerSecond(primaryAttributeID: skill.primaryAttributeID, secondaryAttribute: skill.secondaryAttributeID)
	}
	
	func skillpointsPerSecond(primaryAttributeID: NCDBAttributeID, secondaryAttribute: NCDBAttributeID) -> Double {
		let effectivePrimaryAttribute = effectiveAttributeValue(attributeID: primaryAttributeID)
		let effectiveSecondaryAttribute = effectiveAttributeValue(attributeID: secondaryAttribute)
		return (Double(effectivePrimaryAttribute) + Double(effectiveSecondaryAttribute) / 2.0) / 60.0;

	}
	
	func effectiveAttributeValue(attributeID: NCDBAttributeID) -> Int {
		switch attributeID {
		case .intelligence:
			return intelligence
		case .memory:
			return memory
		case .perception:
			return perception
		case .willpower:
			return willpower
		case .charisma:
			return charisma
		default:
			return 0
		}
	}
}
